wallet/src/de/schildbach/wallet/payments/SendCoinsTaskRunner.kt

```diff
@@ -55,17 +56,63 @@ class SendCoinsTaskRunner @Inject constructor(
         return sendCoins(wallet, sendRequest, scryptIterationsTarget)
     }
 
+    override suspend fun estimateNetworkFee(
+        address: Address,
+        amount: Coin,
+        constrainInputsTo: Address?,
+        emptyWallet: Boolean
+    ): SendPaymentService.TransactionDetails {
+        val wallet = walletApplication.wallet ?: throw RuntimeException("this method can't be used before creating the wallet")
+        var sendRequest = createSendRequest(address, amount, constrainInputsTo, emptyWallet, false)
+        try {
+            val securityGuard = SecurityGuard()
+            val password = securityGuard.retrievePassword()
+            val scryptIterationsTarget = walletApplication.scryptIterationsTarget()
+            val encryptionKey = deriveKey(wallet, password, scryptIterationsTarget)
+
+            sendRequest.aesKey = encryptionKey
+
+            wallet.completeTx(sendRequest)
+            if (checkDust(sendRequest)){
+                sendRequest = createSendRequest(address, amount, constrainInputsTo, emptyWallet)
+                wallet.completeTx(sendRequest)
+            }
```

Comment:
This check for dust with `checkDust` is very important because it helps to prevent an edge case where the transaction will be rejected by the network.

Unfortunately, recreating the exact same `sendRequest` will not resolve the issue.

From `PaymentProtocolViewModel`, we find that there is a difference in the sendRequests that are recreated related to `forceEnsureMinRequiredFee`.  First it is `false` and if this results in dust, then it is set to `true`
```
                var sendRequest = createSendRequest(wallet, false, paymentIntent, signInputs = false, forceEnsureMinRequiredFee = false)

                wallet.completeTx(sendRequest)
                if (checkDust(sendRequest)) {
                    sendRequest = createSendRequest(wallet, false, paymentIntent, signInputs = false, forceEnsureMinRequiredFee = true)
                    wallet.completeTx(sendRequest)
                }
```

This may be sufficient before the second call to `wallet.completeTx(sendRequest)`
```
        sendRequest.ensureMinRequiredFee = forceEnsureMinRequiredFee
        wallet.completeTx(sendRequest)
```

Reply:
For crowdnode when setting up an account, these three transactions were involved (only fees shown).

```
       fee  0.00001008 DASH/kB, 0.00000227 DASH for 225 bytes
       fee  0.00001008 DASH/kB, 0.00000227 DASH for 225 bytes
       fee  0.00001005 DASH/kB, 0.00000376 DASH for 374 bytes
 ```
 
 I was expecting the fees to be 0.00001000 DASH given that `ensureMinRequiredFee` is set to true rather than 0.00000227 DASH or 0.00000376 DASH.
 
 Perhaps somethings on the backend in DashJ have changed over the years and my understanding no longer applies.

Reply:
After further checking, the check for dust may not be required, but let us leave it for now.

Changes in DashJ in the past year or so have made fee calculation more robust. and less prone to errors.